### PR TITLE
Fix Thread.Start while debugging bug on OSX.

### PR DESCRIPTION
**Pull request by mikem8361**
The OSX exception logic is running on a separate thread from the one that the exception happened. The CatchHardwareExceptionHolder::IsEnabled used to check for h/w exception holders assumed it was running on the thread the exception/holders happened not the exception notification thread. Moved the h/w exception holder count to the CPalThread object instead of a TLS thread variable so the OSX exception code can check it given a CPalThread instance.

The main problem was that the stubmgr when creating a thread (for the start notification) put a breakpoint in ThePreStubPatch which is in the coreclr text section and because the h/w exception holder check was broken, it thought the breakpoint wasn't the debugger's and aborted the coreclr process.  The other part of this fix is to put a h/w exception holder around the called to ThePreStubPatch in prestub.cpp.

The stubmgr's delegate invoke subclass used the wrong registers for Linux/OSX to read the "this" and parameter registers. Added the proper ifdefs and registers (ecx -> rdi, rdx -> rsi) for the Unix platforms.

On both Linux and OSX, the h/w exception holder check in the exception routing logic needed to check if the int3/trap is in DBG_DebugBreak so asserts always abort/core dump. Move DBG_DebugBreak to an assembly worker so the start/end address can be used for this check.

### CODE REVIEW

**Comment by mikem8361** [comment]
@janvorli  and @jkotas could you guys review?

File: src/vm/stubmgr.cpp [hunk 1]
```diff
@@ -2269,7 +2269,11 @@ BOOL DelegateInvokeStubManager::TraceManager(Thread *thread, TraceDestination *t
     {
         LOG((LF_CORDB, LL_INFO10000, "DISM::TraceManager: isSingle\n"));
 
-        orDelegate = (DELEGATEREF)ObjectToOBJECTREF((Object*)(size_t)pContext->Rcx);  // The "this" pointer is rcx
+#ifdef UNIX_AMD64_ABI
+        orDelegate = (DELEGATEREF)ObjectToOBJECTREF((Object*)(size_t)pContext->Rdi);  // The "this" pointer is rdi (see unixasmhelpers.S).
```
Comment (jkotas) [inline src/vm/stubmgr.cpp]: This should be call StubManagerHelpers::GetThisPtr helper instead of the ifdef.

File: src/vm/stubmgr.cpp [hunk 2]
```diff
@@ -2293,11 +2297,19 @@ BOOL DelegateInvokeStubManager::TraceManager(Thread *thread, TraceDestination *t
         if (pStub->GetPatchOffset() != 0)
         {
             // This stub has a hidden return buffer argument.
+#ifdef UNIX_AMD64_ABI
+            orDelegate = (DELEGATEREF)ObjectToOBJECTREF((Object*)(size_t)(pContext->Rsi));
```
Comment (jkotas) [inline src/vm/stubmgr.cpp]: StubManagerHelpers::GetSecondArg

File: src/vm/stubmgr.cpp [hunk 3]
```diff
@@ -2293,11 +2297,19 @@ BOOL DelegateInvokeStubManager::TraceManager(Thread *thread, TraceDestination *t
...
+#else
             orDelegate = (DELEGATEREF)ObjectToOBJECTREF((Object*)(size_t)(pContext->Rdx));
+#endif // UNIX_AMD64_ABI
         }
         else
         {
+#ifdef UNIX_AMD64_ABI
+            orDelegate = (DELEGATEREF)ObjectToOBJECTREF((Object*)(size_t)(pContext->Rdi));
+#else
```
Comment (jkotas) [inline src/vm/stubmgr.cpp]: StubManagerHelpers::GetThisPtr

**Comment by janvorli** [comment]
@mikem8361 your description of how exceptions are handled is slightly misleading. The hardware exception handling on OSX actually does happen on the thread that has thrown the exception and not on the thread of the exception monitor. All the exception monitor thread does is to check if the exception should be handled by us at all and if the check is true, inject exception handler invocation and a helper frame into the thread that created the exception and then lets it continue.

But that's just for clarification for others reading this PR, the change itself LGTM.

**Comment by jkotas** [comment]
LGTM

**Comment by mikem8361** [comment]
What you said is what I meant but I didn't say it very clearly.  I'll take your description below and edit the PR text and make it clearer.

Sent from Windows Mail

**Comment by sergiy-k** [comment]
LGTM

**Comment by mikem8361** [comment]
Found a rare occurring issue (#2035) but will merge and fix it later.

**Pull request merged**